File: weather/weather.pl
use Weather::Underground;
use strict;
use warnings;
use utf8;
binmode(STDOUT, ":utf8");
if ($#ARGV == -1) {
  print "Usage: !weather <location>. Try: 'City', 'City, State', 'State', 'State, Country' and 'Country'.\n";
  exit(-1);
}

my $place = join(" ", @ARGV);

my $weather = Weather::Underground->new(
  place => $place,
  debug => 0,
) || print "Error: $@\n";

my $arrayref = $weather->get_weather() || die "Error: $@\n";

foreach (@$arrayref) {
  my $response = "Weather in $_->{'place'}: ";
  my $d;
  my $e;
  my $f;

  $d = $_->{"celsius"};
  if ($d ne "") {
    $response .= "$d"."ºC";
  }

  $d = $_->{"conditions"};
  if ($d ne "") {
    $response .= "/$d; ";
  } else {
    $response .= "; ";
  }
  
  $d = $_->{"wind_direction"};
  if ($d ne "") {
    $response .= "Wind: $d";
  }
  
  $d = $_->{"wind_kilometersperhour"};
  if ($d ne "") {
    $response .= "/$d Km/h; ";
  } else {
    $response .= "; ";
  }
  
  $d = $_->{"pressure"};
  if ($d ne "") {
    $response .= "Pressure: $d; ";
  }
  
  $d = $_->{"humidity"};
  if ($d ne "") {
    $response .= "Humidity: $d%; ";
  }

  $d = $_->{"sunrise"};
  $e = $_->{"sunset"};
  if ($d ne "") {
    $d = substr($d, 0, -4);
    $e = substr($e, 0, -4);
    $response .= "Sunrise/Sunset: $d/$e; ";
  }

  $d = $_->{"moonphase"};
  $e = $_->{"moonrise"};
  $f = $_->{"moonset"};
  if ($d ne "") {
    $e = substr($e, 0, -4);
    $f = substr($f, 0, -4);
    $response .= "Moon Rise/Set: $d $e/$f; ";
  }

  #  while (my ($key, $value) = each %{$_}) {
  #    if (defined $value) {
  #        print "\t$key = $value\n";
  #      } else {
  #        print "\t$key = ?\n";
  #      }
  #  }

  print $response."\n";
}

File: weather/weather.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json
import sys
import os
import urllib2

# ../mylib.py
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from mylib import print_console

city = ""
forecastnum = 0
if (len(sys.argv) == 1):
	print_console("Usage .w <location> [--forecast [1-9]]")
	sys.exit(0)
for x in range(1, len(sys.argv)):
	if x == 1:
		city = sys.argv[x]
	else:
		if sys.argv[x] != "--forecast":
			city = city + "%20" + sys.argv[x]
		else:
			if x == len(sys.argv) - 1:
				forecastnum = 3
			else:
				try:
					forecastnum = int(sys.argv[x + 1])
				except:
					forecastnum = 3
			break

url = 'https://query.yahooapis.com/v1/public/yql?q=select%20*%20from%20weather.forecast%20where%20woeid%20in%20(select%20woeid%20from%20geo.places(1)%20where%20text=%22{}%22)%20and%20u=%22c%22&format=json&env=store://datatables.org/alltableswithkeys'.format(city)

try:
	response = urllib2.urlopen(url, timeout=5)
except urllib2.URLError, e:
		print_console("There was an error: %r" % e)
		sys.exit(0)
values = json.loads(response.read())
if values['query']['results'] is None:
	print_console("No results found")
else:
	channel = values['query']['results']['channel']
	conditions = channel['item']['condition']
	atemosphere = channel['atmosphere']
	astronomy = channel['astronomy']
	wind = channel['wind']
	print_console("%s Temperature: %s Cº  Condition: %s  Humidity: %s%%  Wind: %s km/h  Sunrise/Sunset: %s/%s" % (channel['item']['title'], conditions['temp'], conditions['text'], atemosphere['humidity'], wind['speed'], astronomy['sunrise'], astronomy['sunset'] ) )
	if forecastnum > 0:
		if forecastnum > 9:
			forecastnum = 9
		forecast = channel['item']['forecast']
		forecastprint = ""
		for x in range(1, forecastnum + 1):
			forecastprint = forecastprint + "Forecast for %s: High: %s Low: %s Condition: %s " % (forecast[x]['date'], forecast[x]['high'], forecast[x]['low'], forecast[x]['text'])
		print (forecastprint)
